api/android/api/src/main/jni/nnstreamer-native-api.c

```diff
@@ -635,6 +635,93 @@ nns_get_nnfw_type (jint fw_type, ml_nnfw_type_e * nnfw)
 }
 
+/**
+ * @brief Set additional environment
+ */
+static gboolean
+nns_set_env (JNIEnv * env, jobject context)
+{
+  gboolean ret = TRUE;
+
+#if defined (ENABLE_SNPE)
+  jclass context_class;
+  jmethodID get_application_info_method_id;
+  jobject application_info_object;
+  jfieldID native_Library_Dir_Field_Id;
+  jstring native_library_dir_path;
+  const gchar *native_library_dir_path_str;
+
+  context_class = (*env)->GetObjectClass(env, context);
+  if (!context_class) {
+    nns_loge ("Failed to get context class.");
+    goto failed;
+  }
+
+  get_application_info_method_id = (*env)->GetMethodID (env, context_class,
+      "getApplicationInfo", "()Landroid/content/pm/ApplicationInfo;");
+  if (!get_application_info_method_id) {
+    nns_loge ("Failed to get method ID for `ApplicationInfo()`.");
+    (*env)->DeleteLocalRef (env, context_class);
+    goto failed;
+  }
+
+  application_info_object = (*env)->CallObjectMethod (env, context,
+      get_application_info_method_id);
+  if ((*env)->ExceptionCheck (env)) {
+    (*env)->ExceptionDescribe (env);
+    (*env)->ExceptionClear (env);
+    (*env)->DeleteLocalRef (env, context_class);
+    goto failed;
+  }
+
+  native_Library_Dir_Field_Id = (*env)->GetFieldID (env,
+      (*env)->GetObjectClass (env, application_info_object),
+      "nativeLibraryDir", "Ljava/lang/String;");
+  if (!native_Library_Dir_Field_Id) {
+    nns_loge ("Failed to get field ID for nativeLibraryDir.");
+    (*env)->DeleteLocalRef (env, application_info_object);
+    (*env)->DeleteLocalRef (env, context_class);
+    goto failed;
+  }
+
+  native_library_dir_path = (*env)->GetObjectField (env,
+      application_info_object, native_Library_Dir_Field_Id);
+  if (!native_library_dir_path) {
+    nns_loge ("Failed to get Object Field nativeLibraryDir.");
+    (*env)->DeleteLocalRef (env, application_info_object);
+    (*env)->DeleteLocalRef (env, context_class);
+    goto failed;
+  }
+
+  native_library_dir_path_str = (*env)->GetStringUTFChars (env,
+      native_library_dir_path, NULL);
+  if ((*env)->ExceptionCheck (env)) {
+    (*env)->ExceptionDescribe (env);
+    (*env)->ExceptionClear (env);
+    (*env)->DeleteLocalRef (env, native_library_dir_path);
+    (*env)->DeleteLocalRef (env, application_info_object);
+    (*env)->DeleteLocalRef (env, context_class);
+    goto failed;
+  }
+
+  /* See https://developer.qualcomm.com/docs/snpe/dsp_runtime.html for details */
+  nns_logi ("Set env ADSP_LIBRARY_PATH for snpe DSP/AIP runtime: %s",
+      native_library_dir_path_str);
+  g_setenv ("ADSP_LIBRARY_PATH", native_library_dir_path_str, TRUE);
+
+  (*env)->ReleaseStringUTFChars (env, native_library_dir_path,
+      native_library_dir_path_str);
+  (*env)->DeleteLocalRef (env, native_library_dir_path);
+  (*env)->DeleteLocalRef (env, application_info_object);
+  (*env)->DeleteLocalRef (env, context_class);
+#endif
+
+  return ret;
+
+failed:
+  return FALSE;
```

Comment:
This is just my humble opinion. I think we usually use `goto` statement to reduce some duplicated error handling codes. Unfortunately, this goto statement does not meet such a purpose because it eventually just returns `FALSE`. But, I think there's an opportunity to remove duplicated codes (e.g., `DeleteLocalRef`).

Reply:
Thanks for pointing out! I updated the part. Please check 👀 !